From fe3e47f80b7a9832ae7483d201bd9fb92be2ecd5 Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Wed, 23 Jul 2014 13:27:08 -0400
Subject: [input] synaptics: add a matches_pnp_id helper function

Message-id: <1406122030-5632-13-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 86355
O-Subject: [RHEL7.1 / RHEL7.0.z PATCH 12/14] Input: synaptics - add a matches_pnp_id helper function
Bugzilla: 1093449
RH-Acked-by: Aristeu S. Rozanski F. <arozansk@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Peter Hutterer <peter.hutterer@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1093449
Upstream: since v3.15, backported in stable
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7722437
Test: tested by myself on a t440s and a t410, was extensively tested in RHEL 6.6

commit e2f611029b370bb7a04236215ad4b36aa8cb98cd
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Mon May 19 22:53:23 2014 -0700

    Input: synaptics - add a matches_pnp_id helper function

    This is a preparation patch for simplifying the min/max quirk table.

    Cc: stable@vger.kernel.org
    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/mouse/synaptics.c b/drivers/input/mouse/synaptics.c
index 898e8bb..395ec9c 100644
--- a/drivers/input/mouse/synaptics.c
+++ b/drivers/input/mouse/synaptics.c
@@ -156,6 +156,18 @@ static const char * const topbuttonpad_pnp_ids[] = {
  NULL
 };
 
+static bool matches_pnp_id(struct psmouse *psmouse, const char * const ids[])
+{
+ int i;
+
+ if (!strncmp(psmouse->ps2dev.serio->firmware_id, "PNP:", 4))
+  for (i = 0; ids[i]; i++)
+   if (strstr(psmouse->ps2dev.serio->firmware_id, ids[i]))
+    return true;
+
+ return false;
+}
+
 /*****************************************************************************
  * Synaptics communications functions
  ****************************************************************************/
@@ -1365,17 +1377,8 @@ static void set_input_params(struct psmouse *psmouse,
 
  if (SYN_CAP_CLICKPAD(priv->ext_cap_0c)) {
   __set_bit(INPUT_PROP_BUTTONPAD, dev->propbit);
-  /* See if this buttonpad has a top button area */
-  if (!strncmp(psmouse->ps2dev.serio->firmware_id, "PNP:", 4)) {
-   for (i = 0; topbuttonpad_pnp_ids[i]; i++) {
-    if (strstr(psmouse->ps2dev.serio->firmware_id,
-        topbuttonpad_pnp_ids[i])) {
-     __set_bit(INPUT_PROP_TOPBUTTONPAD,
-        dev->propbit);
-     break;
-    }
-   }
-  }
+  if (matches_pnp_id(psmouse, topbuttonpad_pnp_ids))
+   __set_bit(INPUT_PROP_TOPBUTTONPAD, dev->propbit);
   /* Clickpads report only left button */
   __clear_bit(BTN_RIGHT, dev->keybit);
   __clear_bit(BTN_MIDDLE, dev->keybit);
-- 
1.7.1